### PR TITLE
Subduction cookbook case 2a and 2b

### PR DESCRIPTION
**Pull request by anne-glerum**
This PR adds Case 2a and 2b of the kinematically driven 2d subduction cookbook. The benchmarks by Quinquis et al. requires adding a plugin for the initial temperature distribution and several postprocessing plugins. 

I'll still need to squash the commits and add the Elefant data that is plotted in the diagnostics files, but otherwise this is ready for a review by @cedrict.

### For all pull requests:

* [x] I have followed the [instructions for indenting my code](../blob/main/CONTRIBUTING.md#making-aspect-better).

### For new features/models or changes of existing features:

* [x] I have tested my new feature locally to ensure it is correct.
* [ ] I have [created a testcase](http://www.math.clemson.edu/~heister/manual.pdf#sec%3Awriting_tests) for the new feature/benchmark in the [tests/](../blob/main/tests/) directory.
* [ ] I have added a changelog entry in the [doc/modules/changes](../blob/main/doc/modules/changes) directory that will inform other users of my change.

### CODE REVIEW

File: cookbooks/kinematically_driven_subduction_2d/doc/kinematically_driven_subduction_2d.md
Comment (cedrict) [inline cookbooks/kinematically_driven_subduction_2d/doc/kinematically_driven_subduction_2d.md:108]: may be explain how this initial temperature field is built ? it was a summation of many terms, wasn't it ?
Then the comment about it being 'complex' makes a bit more sense.
Also I am not sure about 'comfortably'.

File: cookbooks/kinematically_driven_subduction_2d/doc/kinematically_driven_subduction_2d.md
Comment (cedrict) [inline cookbooks/kinematically_driven_subduction_2d/doc/kinematically_driven_subduction_2d.md:124]: The thermal expansivity is kept zero in Case 2a so that the thermal evolution...

File: cookbooks/kinematically_driven_subduction_2d/doc/kinematically_driven_subduction_2d.md
Comment (cedrict) [inline cookbooks/kinematically_driven_subduction_2d/doc/kinematically_driven_subduction_2d.md:126]: is assigned to the upper mantle

File: cookbooks/kinematically_driven_subduction_2d/doc/kinematically_driven_subduction_2d.md
Comment (cedrict) [inline cookbooks/kinematically_driven_subduction_2d/doc/kinematically_driven_subduction_2d.md:139]: whole domain over time and the 800 &degC isotherm depth is also tracked.

File: cookbooks/kinematically_driven_subduction_2d/doc/kinematically_driven_subduction_2d.md
Comment (cedrict) [inline cookbooks/kinematically_driven_subduction_2d/doc/kinematically_driven_subduction_2d.md:157]: I may have missed it, but you do not specify what this Elefant code is. I'd just add a citation my rejected paper thieulot:2015, no need to expand on this.Why no SULEC results?

File: cookbooks/kinematically_driven_subduction_2d/doc/kinematically_driven_subduction_2d.md
Comment (cedrict) [inline cookbooks/kinematically_driven_subduction_2d/doc/kinematically_driven_subduction_2d.md:164]: This renders the density (and therefore the Stokes equations) temperature-dependent.

File: cookbooks/kinematically_driven_subduction_2d/doc/kinematically_driven_subduction_2d.md
Comment (cedrict) [inline cookbooks/kinematically_driven_subduction_2d/doc/kinematically_driven_subduction_2d.md:196]: may be add here a cliffhanger? i.e. something like, of course case 2b remains a very simplified model as all compositions are linear viscous. We will address this in the Case 3 models which will be added here in the future.

File: cookbooks/kinematically_driven_subduction_2d/composition_trms_statistics.cc
Comment (cedrict) [inline cookbooks/kinematically_driven_subduction_2d/composition_trms_statistics.cc:2]: in all the files containing this copyright line, there is no unity with regards to the years.

File: cookbooks/kinematically_driven_subduction_2d/doc/kinematically_driven_subduction_2d.md
Comment (cedrict) [inline cookbooks/kinematically_driven_subduction_2d/doc/kinematically_driven_subduction_2d.md:13]: I don't think you do (yet)

**Review by cedrict — CHANGES_REQUESTED** [review]
Thank you for contributing these models. I have a few minor text comments.

File: cookbooks/kinematically_driven_subduction_2d/doc/kinematically_driven_subduction_2d.md
Comment (anne-glerum) [reply cookbooks/kinematically_driven_subduction_2d/doc/kinematically_driven_subduction_2d.md:13]: Indeed we still don't :D. I wrote that in the Case 1 PR though, so if it was okay then, I guess it's still okay?

File: cookbooks/kinematically_driven_subduction_2d/doc/kinematically_driven_subduction_2d.md
Comment (anne-glerum) [reply cookbooks/kinematically_driven_subduction_2d/doc/kinematically_driven_subduction_2d.md:157]: Good point! I do have some SULEC results, but since I didn't ask for permission to use them, I didn't add them.

File: cookbooks/kinematically_driven_subduction_2d/doc/kinematically_driven_subduction_2d.md
Comment (cedrict) [reply cookbooks/kinematically_driven_subduction_2d/doc/kinematically_driven_subduction_2d.md:13]: sure :)

**Comment by cedrict** [comment]
@naliboff I had a look at the text/figures, but I feel somebody more experienced should look at the code

File: cookbooks/kinematically_driven_subduction_2d/isotherm_depth.cc
Comment (naliboff) [inline cookbooks/kinematically_driven_subduction_2d/isotherm_depth.cc:76]: Add a space above this line

File: cookbooks/kinematically_driven_subduction_2d/isotherm_depth.cc
Comment (naliboff) [inline cookbooks/kinematically_driven_subduction_2d/isotherm_depth.cc:82]:
```suggestion
      // also make sure that the other columns filled by this object
```

File: cookbooks/kinematically_driven_subduction_2d/isotherm_depth.cc
Comment (naliboff) [inline cookbooks/kinematically_driven_subduction_2d/isotherm_depth.cc:114]:
```suggestion
                            "which maximum depth is tracked with. Unit: [K].");
```

File: cookbooks/kinematically_driven_subduction_2d/isotherm_depth.h
Comment (naliboff) [inline cookbooks/kinematically_driven_subduction_2d/isotherm_depth.h:2]:
```suggestion
  Copyright (C) 2024 by the authors of the ASPECT code.
```

File: cookbooks/kinematically_driven_subduction_2d/isotherm_depth.h
Comment (naliboff) [inline cookbooks/kinematically_driven_subduction_2d/isotherm_depth.h:43]: Is this object actually evaluating the solution, or just storing the results of the evaluation? From the .cc code, it looks like it is just used to store the values?

File: cookbooks/kinematically_driven_subduction_2d/kinematically_driven_subduction_2d_case2a.prm
Comment (naliboff) [inline cookbooks/kinematically_driven_subduction_2d/kinematically_driven_subduction_2d_case2a.prm:12]:
```suggestion
# High resolution is required to sufficiently resolve the weak zone
```

File: cookbooks/kinematically_driven_subduction_2d/kinematically_driven_subduction_2d_case2a.prm
Comment (naliboff) [inline cookbooks/kinematically_driven_subduction_2d/kinematically_driven_subduction_2d_case2a.prm:13]:
```suggestion
# between the subducting and overriding plate, and consequently
```

File: cookbooks/kinematically_driven_subduction_2d/kinematically_driven_subduction_2d_case2a.prm
Comment (naliboff) [inline cookbooks/kinematically_driven_subduction_2d/kinematically_driven_subduction_2d_case2a.prm:14]: Can you provide a range of MPI processes here? Even better would be to say something like `.. overriding plate, which results in a model size of approximately X degrees of freedom. This requires on the order of X-Y processors to run efficiently over X-Y hours.`

File: cookbooks/kinematically_driven_subduction_2d/kinematically_driven_subduction_2d_case2b.prm
Comment (naliboff) [inline cookbooks/kinematically_driven_subduction_2d/kinematically_driven_subduction_2d_case2b.prm:14]: I think it would be ok to replace all of the text above with:
`A modification of kinematically driven 2d subduction cookbook case 2a to include temperature-dependent density.`

File: cookbooks/kinematically_driven_subduction_2d/kinematically_driven_subduction_2d_case2b.prm
Comment (naliboff) [inline cookbooks/kinematically_driven_subduction_2d/kinematically_driven_subduction_2d_case2b.prm:29]: If any of the parameters below are reused from case 2a (Thermal conductivities, Heat Capacities), consider removing them. I would always add spaces between different sets of parameters to improve the readability.

File: cookbooks/kinematically_driven_subduction_2d/subduction_plate_cooling.cc
Comment (naliboff) [inline cookbooks/kinematically_driven_subduction_2d/subduction_plate_cooling.cc:2]:
```suggestion
  Copyright (C) 2024 by the authors of the ASPECT code.
```

File: cookbooks/kinematically_driven_subduction_2d/subduction_plate_cooling.cc
Comment (naliboff) [inline cookbooks/kinematically_driven_subduction_2d/subduction_plate_cooling.cc:36]:
```suggestion
      // box. Throw an assert if this is not the case.
```

**Review by naliboff — COMMENTED** [review]
@anne-glerum (and @cedrict) - Thanks for contributing these cookbooks! I have a number of comments throughout, but all comments fall in the category of minor cleanup by and large. In some place I've suggested modifications to the PRM files that can be taken care of in follow up PRs, so I will leave it up to you whether you would like to implement them in this PR. Thanks again!

File: cookbooks/kinematically_driven_subduction_2d/doc/Case2a_temperatures.prm
Comment (anne-glerum) [reply cookbooks/kinematically_driven_subduction_2d/doc/Case2a_temperatures.prm:2]: I haven't tried that, but the student who will continue with Case 3 and 4 will probably give it a try.

File: cookbooks/kinematically_driven_subduction_2d/doc/kinematically_driven_subduction_2d.md
Comment (anne-glerum) [reply cookbooks/kinematically_driven_subduction_2d/doc/kinematically_driven_subduction_2d.md:165]: I've added a derivation in the prm snippet now.

File: cookbooks/kinematically_driven_subduction_2d/doc/Case1_compositions.prm
Comment (anne-glerum) [reply cookbooks/kinematically_driven_subduction_2d/doc/Case1_compositions.prm:16]: I think it's okay as the abbreviations match with the definitions of the abbreviations in the text.

File: cookbooks/kinematically_driven_subduction_2d/doc/Case1_materialmodel.prm
Comment (anne-glerum) [reply cookbooks/kinematically_driven_subduction_2d/doc/Case1_materialmodel.prm:11]: Oh we can do that, that's great!

Comment (anne-glerum) [reply cookbooks/kinematically_driven_subduction_2d/doc/Case1_materialmodel.prm:11]: I don't think we can do that yet, except when using phase changes. Would you have a working example?

Comment (naliboff) [reply cookbooks/kinematically_driven_subduction_2d/doc/Case1_materialmodel.prm:11]: @anne-glerum - My mistake, I thought this was enabled with the changes you and Bob implemented.

File: cookbooks/kinematically_driven_subduction_2d/subduction_plate_cooling.cc
Comment (anne-glerum) [reply cookbooks/kinematically_driven_subduction_2d/subduction_plate_cooling.cc:41]: That would be easier for the user, but since this plugin is just for the cookbook reproducing the benchmark setup, and there already is a plate cooling option in the adiabatic initial temperature plugin, I'd prefer not to make these variables parsable from the input file. Would that be okay?

**Comment by anne-glerum** [comment]
Thanks for the review John! I've addressed and/or answered your points.

**Comment by gassmoeller** [comment]
Please squash the commits, you have 98 on this branch.

**Review by naliboff — COMMENTED** [review]
@anne-glerum - I think it might be nice to eventually move some of the hard code parameters for the model into user parameters, but that can be a follow up PR if someone wishes to do so down the line. 

For now, I think the only thing left to do is add test cases for some or all of the cookbook prms to make sure they run. After that, I think we should merge and do follow up PRs as necessary for minor cleanup or additional features.

**Comment by anne-glerum** [comment]
This PR should be done now, I've just updated the test results.

**Comment by gassmoeller** [comment]
Please dont merge this yet, I would like to read through the .cc files and I dont have the time right now. I am sure the benchmarks are fine, but I saw at least on code line that may produce compiler issues on some systems.

File: cookbooks/kinematically_driven_subduction_2d/composition_trms_statistics.h
Comment (gassmoeller) [inline cookbooks/kinematically_driven_subduction_2d/composition_trms_statistics.h:24]: Double underscores are reserved for compiler purposes. We switched a while ago, but this branch was created before that time.
```suggestion
#ifndef _aspect_postprocess_composition_trms_statistics_h
#define _aspect_postprocess_composition_trms_statistics_h

```

File: cookbooks/kinematically_driven_subduction_2d/isotherm_depth.h
Comment (gassmoeller) [inline cookbooks/kinematically_driven_subduction_2d/isotherm_depth.h:22]: double underscores

**Review by gassmoeller — APPROVED** [review]
Let's merge this before we forget. I will prepare a follow-up PR with minor improvements (nothing in this PR is wrong).

**Comment by anne-glerum** [comment]
Thank you Rene!

**Pull request merged**
